### ISSUE TITLE
Backprop issue I can't figure out

### ISSUE THREAD

**Issue by jekbradbury** [pytorch/pytorch#619]
Here's a straightforward one-layer bidirectional LSTM for sentiment classification -- but it only trains with `self.l_out(hn2)` while the loss stays above 0.6 with `self.l_out(hn1)` (yet the `assert` is never violated). The same is true if I used fixed embeddings with `.detach()` or pretrained GloVe embeddings.
```python
import torch
from torch import nn
from torch import optim
from torch.autograd import Variable

from torchtext import data
from torchtext import datasets

batch_size = 64

TEXT = data.Field()
LABEL = data.Field(sequential=False)

train, valid, test = datasets.SST.splits(
    TEXT, LABEL, fine_grained=False, train_subtrees=True,
    filter_pred=lambda ex: ex.label != 'neutral')

TEXT.build_vocab(train)
LABEL.build_vocab(train)

train_iter = data.BucketIterator(train, batch_size=batch_size, device=0)

num_classes = len(LABEL.vocab)
num_input_symbols = len(TEXT.vocab)
rnn_size = 100

class lstm_bi(nn.Module):
    def __init__(self):
        super().__init__()
        self.embedding = nn.Embedding(num_input_symbols, rnn_size)
        self.rnn = nn.LSTM(rnn_size, rnn_size, 1, bidirectional=True)
        self.l_out = nn.Linear(rnn_size * 2, num_classes)

    def forward(self, x):
        embeds = self.embedding(x)
        _batch_size = embeds.size(1)
        h0 = Variable(torch.zeros(1 * 2, _batch_size, rnn_size).cuda())
        c0 = Variable(torch.zeros(1 * 2, _batch_size, rnn_size).cuda())
        _, (hn, _) = self.rnn(embeds, (h0, c0))
        hn1 = hn.transpose(0, 1).contiguous().view(_batch_size, -1)
        hn2 = torch.cat([hn[0], hn[1]], 1)
        diff = (hn1 - hn2).norm().data[0]
        assert diff == 0
        return self.l_out(hn2)

net = lstm_bi()
net.cuda()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(net.parameters())

for batch in train_iter:
    outputs = net(batch.text)
    cost = criterion(outputs, batch.label)
    print(cost.data[0])
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
```
If I use just the formulas for `hn1` and `hn2` in isolation, `.backward()` gives the same results for `hn.grad` for each; the only gradient difference I've found is that if I iterate through the parameters after `cost.backward()`:
```python
for i, param in enumerate(net.parameters()):
   print(i, param.grad.data.norm())
```
some of the LSTM gradients are zero when using `hn2`. But none of them should be zero...and `hn2` is the one that works...

**Comment by apaszke** [comment]
Weird, maybe it's a cuDNN issue? After 1600 iterations on CPU I got 0.27 cost with `hn1` and 0.28 with `hn2`.

**Comment by apaszke** [comment]
Yup, confirmed to be a cuDNN problem:

|       | CPU  | GPU  | cuDNN |
|-------|------|------|-------|
| `hn1` | 0.27 | 0.26 | 0.67  |
| `hn2` | 0.28 | 0.28 | 0.27  |

**Comment by apaszke** [comment]
I just sent a PR with a fix #620.

**Comment by apaszke** [comment]
Merged. This should be fixed now.
